Implementing residuals

AlexBigO opened the pull request.

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+            The residuals
+        """
+        residuals = []
+        chi2 = self.get_chi2_ndf()*self._ndf_
```
```suggestion
        chi2 = self.get_chi2()
```

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+    def get_raw_residuals(self, idx2=0):
+        """
+        Get the raw residuals (data_value - bkg_model_value) for all bins
+
```
The documentation about the input parameter `idx2` is missing

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+        norm = self._data_handler_.get_norm()
+        data_values = binned_data.values()/norm
+        # access model predicted values for background
+        background_pdf_binned_ = zfit.pdf.BinnedFromUnbinnedPDF(self._background_pdf_[idx2],
+                                                                self._data_handler_.get_obs())
```
this only works if you have a single background, so it should be extended and loop over all the defined bkg pdfs

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+        background_pdf_binned_ = zfit.pdf.BinnedFromUnbinnedPDF(self._background_pdf_[idx2],
+                                                                self._data_handler_.get_obs())
+        signal_fracs, _, _, _ = self.__get_all_fracs()
+        bkg_frac = 1 - signal_fracs[idx2]
```
same here: this works only if you have a single background and a single signal pdf

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+
+        residuals, variances = self.get_raw_residuals()
+        # residuals into binned data
+        residuals_binned_data = zfit.data.BinnedData.from_tensor(space=obs,
+                                                                values=residuals,
+                                                                variances=variances)
```
```suggestion
        residuals_binned_data = zfit.data.BinnedData.from_tensor(space=obs,
                                                                 values=residuals,
                                                                 variances=variances)
```

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+    def plot_raw_residuals(self, **kwargs):
+        """
+        Plot the raw residuals
+
```
Also here the documentation about the `kwargs` is missing

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+            signal_fracs.append(1.)
+
+        # draw signals
+        base_sgn_cmap = plt.cm.get_cmap('viridis', len(signal_funcs) * 4)
+        sgn_cmap = ListedColormap(base_sgn_cmap(np.linspace(0.4, 0.65, len(signal_funcs))))
```
To not duplicate code, this could become a data member

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+            The standardized residuals
+        """
+        residuals, variances = [], []
+        chi2 = self.get_chi2_ndf()*self._ndf_
```
```suggestion
        chi2 = self.get_chi2()
```

fgrosa commented on flarefly/data_handler.py:
```diff
@@ -66,13 +67,13 @@ def __init__(self, data=None, var_name='', limits=None, use_zfit=True, **kwargs)
                             self._limits_[1],
                             name="xaxis"
                         )
-                        self._obs_ = zfit.Space("xaxis", binning=binning)
+                        self._obs_ = zfit.Space("xaxis", binning=self._binning_)
```
I think that there is no need to add the binning as datamember, since you can retrieve the binning from `self._obs_` in the `F2Fitter` class as `self._data_handler_.get_obs().binning`

fgrosa commented on flarefly/fitter.py:
```diff
@@ -736,6 +736,193 @@ def get_chi2_ndf(self):
+            bins = self._data_handler_.get_nbins()
+            bin_sigma = (limits[1] - limits[0]) / bins
+
+        # line at 0
+        plt.scatter(bin_center, [0]*bins, marker="_", s=1, color='xkcd:blue')
+        # line at -3 sigma
+        plt.scatter(bin_center, [-3]*bins, marker="_", s=1, color='xkcd:red')
+        # line at 3 sigma
+        plt.scatter(bin_center, [3]*bins, marker="_", s=1, color='xkcd:red')
```
```suggestion
        # line at 0
        plt.plot([bin_center[0], bin_center[1]], [0., 0.], lw=2, color='xkcd:blue')
        # line at -3 sigma
        plt.plot([bin_center[0], bin_center[1]], [-3., -3.], lw=2, color='xkcd:red')
        # line at 3 sigma
        plt.plot([bin_center[0], bin_center[1]], [3., 3.], lw=2, color='xkcd:red')
```

fgrosa reviewed (changes requested):
Ciao @AlexBigO, thanks a lot for the implementation! The only major comment I have is that the `get_raw_residuals` function works only in case of 1 signal and 1 background (it should subtract all the backgrounds and not subtract all the signals).

AlexBigO commented:
Hey @fgrosa , thanks for the comments. I tried to correct what I did, is this new version okay?

AlexBigO commented:
I removed some unused variables and also added `FATAL` Loggers in case one calls the `residuals` functions while data are unbinned (as it was not implemented).

fgrosa commented:
> I removed some unused variables and also added `FATAL` Loggers in case one calls the `residuals` functions while data are unbinned (as it was not implemented).

It looks good now to me thanks a lot! 😃 You just have to solve a couple of format issues to make pylint happy, then we can merge it

AlexBigO commented:
I am indeed struggling with pylint ^^
Do you have an idea of how to solve these issues?
```
************* Module flarefly.fitter
flarefly/fitter.py:17:0: R0904: Too many public methods (24/20) (too-many-public-methods)
```
For the `Too many pucliv methods`, apart from moving some methods to another `.py` I don't see what could be done. Even adding a `plot_utils.py` would only remove 3 methods from `fitter.py`. Maybe a `qc_utils.py` could also be created, we would store all the chi2, residuals (and future hypothesis tests) in it.

fgrosa commented:
> I am indeed struggling with pylint ^^ Do you have an idea of how to solve this issue?
> 
> ```
> ************* Module flarefly.fitter
> flarefly/fitter.py:17:0: R0904: Too many public methods (24/20) (too-many-public-methods)
> ```
> 
> For the `Too many pucliv methods`, apart from moving some methods to another `.py` I don't see what could be done. Even adding a `plot_utils.py` would only remove 3 methods from `fitter.py`. Maybe a `qc_utils.py` could also be created, we would store all the chi2, residuals (and future hypothesis tests) in it.

I think that for the moment you can just add 
```
#pylint: disable=too-many-public-methods
```

Then, as next improvement, we can indeed create a module for the plotting utilities and maybe we can set the getters as properties (@property, see e.g. https://www.programiz.com/python-programming/property) so to reduce the number of public methods. However for the time being I would not move the methods for the computation of the residuals and chi2 outside (we might want to do that if we add more sophisticated statistics tests, but for the time being I would say that it's more convenient to have them in the fitter class)
